pkg/event-manager/drivers/docker_backend.go

```diff
@@ -0,0 +1,170 @@
+		}
+		for key, val := range secret.Secrets {
+			secrets[key] = val
+		}
+	}
+	return secrets, nil
+}
+
+// Deploy deploys driver
+func (d *dockerBackend) Deploy(ctx context.Context, driver *entities.Driver) error {
+	log.Infof("Docker backend: deploying driver %v", driver.Name)
+
+	// get driver secrets
+	secrets, err := d.getDriverSecrets(ctx, driver)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("err getting secrets of driver %s", driver.Image))
+	}
+
+	rc, err := d.dockerClient.ImagePull(ctx, driver.Image, types.ImagePullOptions{})
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("err pulling image %s", driver.Image))
+	}
+	defer rc.Close()
+	io.Copy(os.Stdout, rc)
+
+	return utils.Backoff(time.Duration(d.DeployTimeout)*time.Second, func() error {
+		created, err := d.dockerClient.ContainerCreate(ctx, &container.Config{
+			Image: driver.Image,
+			Env:   bindEnv(secrets),
+			Cmd:   buildArgs(driver.Config),
+			Labels: map[string]string{
+				labelEventDriverID: driver.ID,
+			},
+		}, nil, nil, "")
+		if err != nil {
+			return errors.Wrap(err, "error creating container")
+		}
+
+		if err := d.dockerClient.ContainerStart(ctx, created.ID, types.ContainerStartOptions{}); err != nil {
+			return errors.Wrap(err, "error starting container")
+		}
+		return nil
+	})
+}
+
+// Expose exposes driver
+func (d *dockerBackend) Expose(ctx context.Context, driver *entities.Driver) error {
+	log.Infof("Docker backend: exposing driver %v", driver.Name)
+	// TODO: expose
+	return nil
+}
+
+// Update updates driver
+func (d *dockerBackend) Update(ctx context.Context, driver *entities.Driver) error {
+	opts := filters.NewArgs()
+	opts.Add("label", labelEventDriverID+"="+driver.ID)
+	containers, err := d.dockerClient.ContainerList(ctx, types.ContainerListOptions{
+		Filters: opts,
+		All:     true,
+	})
+	if len(containers) != 1 || err != nil {
+		return errors.Wrap(err, "error getting container while updating event driver")
+	}
+
+	// Update the driver in UPDATING status
+	// docker doesn't support change env/args of running container,
+	// will first delete then create
+	if driver.Status == entitystore.StatusUPDATING {
+		if err := d.Delete(ctx, driver); err != nil {
+			return errors.Wrap(err, "error deleting container before update")
+		}
+		if err := d.Deploy(ctx, driver); err != nil {
+			return errors.Wrap(err, "error deploying container during update")
+		}
+		log.Infof("driver %s updated", driver.Name)
+	}
+	return nil
+}
+
+// Delete deletes driver
+func (d *dockerBackend) Delete(ctx context.Context, driver *entities.Driver) error {
+	log.Infof("Docker backend: deleting driver %v", driver.Name)
+
+	opts := filters.NewArgs()
+	opts.Add("label", labelEventDriverID+"="+driver.ID)
+	containers, err := d.dockerClient.ContainerList(ctx, types.ContainerListOptions{
+		Filters: opts,
+		All:     true,
+	})
+	if len(containers) != 1 || err != nil {
+		return errors.Wrap(err, "error getting container while deleting event driver")
+	}
```

Comment:
It might be useful to have a private `delete` method that just takes a container so we don't have to repeat this look up logic during an update.

Reply:
I removed the container look up in Update(). Since we only need to deal with container in delete.